From 96083a50ddfa1ddbc9050940c4ce4a86568f49aa Mon Sep 17 00:00:00 2001
From: Thiago Macieira <thiago.macieira@intel.com>
Date: Tue, 29 Oct 2019 20:14:31 -0700
Subject: [llvm] Build tablegen component as a shared library

The LLVMTableGen component is a special case that is excluded from libLLVM and
normally is only built as static library and linked into llvm-tblgen.

We need to have it as a shared library to be available for other projects such
as ldc.

This patch makes it even more special and forces it to be build and installed
as separate shared library.

Source: https://build.opensuse.org/package/view_file/devel:tools:compiler/llvm10/llvm_build_tablegen_component_as_shared_library.patch?expand=1
---
 llvm/cmake/modules/AddLLVM.cmake  | 2 +-
 llvm/cmake/modules/TableGen.cmake | 3 ++-
 llvm/lib/TableGen/CMakeLists.txt  | 6 ++++++
 3 files changed, 9 insertions(+), 2 deletions(-)

diff --git llvm/cmake/modules/AddLLVM.cmake llvm/cmake/modules/AddLLVM.cmake
index 74a07117918a..6bf243be0587 100644
--- llvm/cmake/modules/AddLLVM.cmake
+++ llvm/cmake/modules/AddLLVM.cmake
@@ -616,7 +616,7 @@ function(llvm_add_library name)
   if(ARG_MODULE AND LLVM_EXPORT_SYMBOLS_FOR_PLUGINS AND ARG_PLUGIN_TOOL AND (WIN32 OR CYGWIN))
     # On DLL platforms symbols are imported from the tool by linking against it.
     set(llvm_libs ${ARG_PLUGIN_TOOL})
-  elseif (NOT ARG_COMPONENT_LIB)
+  elseif (NOT ARG_COMPONENT_LIB  OR (${name} STREQUAL "LLVMTableGen" AND LLVM_BUILD_LLVM_DYLIB))
     if (LLVM_LINK_LLVM_DYLIB AND NOT ARG_DISABLE_LLVM_LINK_LLVM_DYLIB)
       set(llvm_libs LLVM)
     else()
diff --git llvm/cmake/modules/TableGen.cmake llvm/cmake/modules/TableGen.cmake
index 73c1e96d3d9a..6290416044d2 100644
--- llvm/cmake/modules/TableGen.cmake
+++ llvm/cmake/modules/TableGen.cmake
@@ -145,7 +145,8 @@ macro(add_tablegen target project)
     set(LLVM_ENABLE_OBJLIB ON)
   endif()
 
-  add_llvm_executable(${target} DISABLE_LLVM_LINK_LLVM_DYLIB ${ARGN})
+  add_llvm_executable(${target} ${ARGN})
+  target_link_libraries(${target} PRIVATE LLVMTableGen)
   set(LLVM_LINK_COMPONENTS ${${target}_OLD_LLVM_LINK_COMPONENTS})
 
   set(${project}_TABLEGEN "${target}" CACHE
diff --git llvm/lib/TableGen/CMakeLists.txt llvm/lib/TableGen/CMakeLists.txt
index 1ece81459520..6a42203046cc 100644
--- llvm/lib/TableGen/CMakeLists.txt
+++ llvm/lib/TableGen/CMakeLists.txt
@@ -1,3 +1,7 @@
+if (LLVM_BUILD_LLVM_DYLIB)
+    set(MAYBE_SHARED SHARED)
+endif()
+
 add_llvm_component_library(LLVMTableGen
   Error.cpp
   JSONBackend.cpp
@@ -9,6 +13,8 @@ add_llvm_component_library(LLVMTableGen
   TGLexer.cpp
   TGParser.cpp
 
+  ${MAYBE_SHARED}
+
   ADDITIONAL_HEADER_DIRS
   ${LLVM_MAIN_INCLUDE_DIR}/llvm/TableGen
   )
-- 
2.31.1

